util/genrest/resttools/systemparam_test.go

```diff
@@ -0,0 +1,259 @@
+			wantInt:     true,
+		},
+		{
+			queryString: "%24alt=json%3Benum-encoding%3Dint",
+			wantInt:     true,
+		},
+
+		// system param+query params in front
+		{
+			queryString: "foo=bar&alt=json",
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+		{
+			queryString: "foo=bar&$alt=json",
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+		{
+			queryString: "foo=bar&%24alt=json",
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+		{
+			queryString: "foo=bar&alt=json%3Benum-encoding=int",
+			wantInt:     true,
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+		{
+			queryString: "foo=bar&$alt=json%3Benum-encoding=int",
+			wantInt:     true,
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+		{
+			queryString: "foo=bar&%24alt=json%3Benum-encoding=int",
+			wantInt:     true,
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+
+		// system param+query params in rear
+		{
+			queryString: "alt=json&foo=bar",
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+		{
+			queryString: "$alt=json&foo=bar",
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+		{
+			queryString: "%24alt=json&foo=bar",
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+		{
+			queryString: "alt=json%3Benum-encoding=int&foo=bar",
+			wantInt:     true,
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+		{
+			queryString: "$alt=json%3Benum-encoding=int&foo=bar",
+			wantInt:     true,
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+		{
+			queryString: "%24alt=json%3Benum-encoding=int&foo=bar",
+			wantInt:     true,
+			wantParams: map[string][]string{
+				"foo": {"bar"},
+			},
+		},
+
+		// system param errors
+		{
+			queryString: "$alt=foo",
+			wantError:   true,
+		},
+		{
+			queryString: "$alt",
+			wantError:   true,
+		},
+		{
+			queryString: "$alt=JSON",
+			wantError:   true,
+		},
+		{
+			queryString: "$alt=json%3Benum-encoding=string",
+			wantError:   true,
+		},
+		{
+			queryString: "$alt=json;enum-encoding=int", // unencoded semicolon
+			wantError:   true,
+		},
+		{
+			queryString: "$alt=json%3Benum-encoding=INT",
+			wantError:   true,
+		},
+		{
+			queryString: "foo&$alt=json&bar&alt=json", // repeated
+			wantError:   true,
+		},
+	} {
+		label := fmt.Sprintf("[%2d %q]", idx, testCase.queryString)
+
+		systemParams, queryParams, err := processQueryString(testCase.queryString)
+
+		if got, want := (err != nil), testCase.wantError; got != want {
+			t.Errorf("%s: error condition not met: want error: %v, got error:%v", label, testCase.wantError, err)
+			continue
+		}
+		if err != nil {
+			continue
+		}
+
+		wantParams := testCase.wantParams
+		if wantParams == nil {
+			wantParams = map[string][]string{}
+		}
+		if got, want := queryParams, wantParams; !reflect.DeepEqual(got, want) {
```

Comment:
Please use [github.com/google/go-cmp](https://pkg.go.dev/github.com/google/go-cmp/cmp) instead of reflect.DeepEqual. There should be other examples of cmp use in this repo.

Reply:
Done